src/mozanalysis/frequentist_stats/sample_size.py

```diff
@@ -500,3 +504,584 @@ def _for_interval_sample_size_calculation(i):
+        self.sample_rate = 0 if sample_rate >= 1 else sample_rate
+
+        self.total_historical_period_days = (
+            n_days_launch_after_version_release
+            + n_days_enrollment
+            + n_days_observation_max
+        )
+        # Will be set after pulling data
+        self._population_size = None
+
+        # Sets self.min_major_version, self.min_major_version_date
+        self._find_latest_feasible_historical_version()
+        self.enrollment_start_date = (
+            date.fromisoformat(self.min_major_version_date)
+            + timedelta(days=self.n_days_launch_after_version_release)
+        ).isoformat()
+
+        print(
+            "\nTotal historical time period will include:",
+            f"\n- {n_days_launch_after_version_release} days wait after new version release to launch",
+            f"\n- {n_days_enrollment} days enrollment",
+            f"\n- up to {n_days_observation_max} days observation.",
+            f"\nTo accomodate this, the historical period will start from version {self.min_major_version}",
+            f"released on {self.min_major_version_date}.",
+            f"Enrollment will start on {self.enrollment_start_date}.",
+        )
+        if self.sample_rate:
+            print(
+                f"Target population will be sampled at a rate of {self.sample_rate:.0%}."
+            )
+
+        # Update segment definitions to use this version and any sampling.
+        self._update_targets_for_historical_period()
+
+    def _find_latest_feasible_historical_version(self) -> None:
+        """Find latest version matching our timeline."""
+        # Allow a gap of 2 days to be on the safe side.
+        latest_feasible_date = date.today() - timedelta(days=2)
+        if self.end_date:
+            end_date = date.fromisoformat(self.end_date)
+        else:
+            end_date = latest_feasible_date
+        if end_date > latest_feasible_date:
+            print(
+                f"\nRequested end date {end_date.isoformat()} is later than",
+                f"the latest feasible date {latest_feasible_date.isoformat()}.",
+                f"Using {latest_feasible_date.isoformat()} as the end date.",
+            )
+            end_date = latest_feasible_date
+        start_date_str = (
+            end_date - timedelta(days=self.total_historical_period_days)
+        ).isoformat()
+
+        # Find the most recent version released prior the desired start date.
+        major_releases = get_firefox_release_dates()
+        self.min_major_version = max(
+            [v for v, d in major_releases.items() if d <= start_date_str]
+        )
+        self.min_major_version_date = major_releases[self.min_major_version]
+
+    def _update_targets_for_historical_period(self) -> None:
+        """Add an additional Segment to the target list to filter on version and sample ID."""
+        historical_filter = (
+            f"(browser_version_info.major_version >= {self.min_major_version})"
```

Comment:
This'll have to be updated for the variety of browser version field names in different data sources, maybe by allowing users to pass a SQL snippet to extract the version from their source.